File: ios/FaceImageProcessor.swift
import ExpoModulesCore
import ImageIO
import UIKit
import UniformTypeIdentifiers
import Vision

internal struct CropOptions: Record {
  @Field var padding: Double = 0.2
  @Field var square: Bool = false
  @Field var outputSize: OutputSize?
  @Field var format: String = "jpeg"
  @Field var quality: Double = 0.9
}

internal struct OutputSize: Record {
  @Field var width: Double = 0
  @Field var height: Double = 0
}

internal struct RectRecord: Record {
  @Field var x: Double = 0
  @Field var y: Double = 0
  @Field var width: Double = 0
  @Field var height: Double = 0

  var cgRect: CGRect { CGRect(x: x, y: y, width: width, height: height) }
}

internal struct PointRecord: Record {
  @Field var x: Double = 0
  @Field var y: Double = 0

  var cgPoint: CGPoint { CGPoint(x: x, y: y) }
}

internal enum FaceImageProcessor {
  static var outputDirectory: URL {
    let caches = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
      ?? FileManager.default.temporaryDirectory
    return caches.appendingPathComponent("face-vision", isDirectory: true)
  }

  // MARK: - Operations

  static func detectFaces(uri: String, options: DetectionOptions) throws -> [String: Any] {
    return try autoreleasepool {
      let start = CACurrentMediaTime()
      let url = try fileURL(from: uri)
      let source = try imageSource(url)
      let size = try uprightSize(source, url)

      // Vision has no fast/accurate switch for faces, so the mode only picks the input size.
      let maxDimension: Int?
      switch options.performanceMode {
      case "fast": maxDimension = 640
      case "accurate": maxDimension = nil
      default: maxDimension = 1280
      }
      let image = try decodeUpright(source, url, maxDimension: maxDimension)
      let handler = VNImageRequestHandler(cgImage: image, orientation: .up, options: [:])
      let faces = try FaceDetector.detect(handler: handler, options: options, size: size)

      return [
        "faces": faces.map(\.dictionary),
        "image": ["width": Double(size.width), "height": Double(size.height), "uri": url.absoluteString],
        "processingTime": (CACurrentMediaTime() - start) * 1000
      ]
    }
  }

  static func analyzeRegion(uri: String, rect: CGRect?) throws -> [String: Any] {
    return try autoreleasepool {
      let url = try fileURL(from: uri)
      let image = try decodeUpright(try imageSource(url), url, maxDimension: nil)
      let full = CGRect(x: 0, y: 0, width: image.width, height: image.height)
      guard let r = RegionStatsCalculator.region(rect ?? full, width: image.width, height: image.height),
        let cropped = image.cropping(to: CGRect(x: r.x, y: r.y, width: r.width, height: r.height)) else {
        throw FaceVisionException(.invalidArgument, "The region does not overlap the image.")
      }

      // ponytail: expands the whole region 1:1 so sampling matches Android exactly; a full
      // 12MP frame is ~48MB here, decode only the sampled rows if that ever matters.
      let rowBytes = r.width * 4
      var rgba = [UInt8](repeating: 0, count: rowBytes * r.height)
      let drawn = rgba.withUnsafeMutableBytes { buffer -> Bool in
        guard let space = CGColorSpace(name: CGColorSpace.sRGB), let context = CGContext(
          data: buffer.baseAddress,
          width: r.width,
          height: r.height,
          bitsPerComponent: 8,
          bytesPerRow: rowBytes,
          space: space,
          bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue
        ) else {
          return false
        }
        context.draw(cropped, in: CGRect(x: 0, y: 0, width: r.width, height: r.height))
        return true
      }
      guard drawn else {
        throw FaceVisionException(.invalidImage, "Could not read the image pixels.")
      }

      return RegionStatsCalculator.stats(r) { x, y in
        let i = (y - r.y) * rowBytes + (x - r.x) * 4
        return 0.299 * Double(rgba[i]) + 0.587 * Double(rgba[i + 1]) + 0.114 * Double(rgba[i + 2])
      }
    }
  }

  static func cropFace(uri: String, rect: CGRect, options: CropOptions) throws -> [String: Any] {
    return try autoreleasepool {
      let url = try fileURL(from: uri)
      let image = try decodeUpright(try imageSource(url), url, maxDimension: nil)
      return try render(image, face: rect, rotation: 0, options: options)
    }
  }

  static func alignFace(uri: String, rect: CGRect, leftEye: CGPoint, rightEye: CGPoint, options: CropOptions) throws -> [String: Any] {
    let dx = leftEye.x - rightEye.x
    let dy = leftEye.y - rightEye.y
    guard dx.isFinite, dy.isFinite, (dx * dx + dy * dy).squareRoot() >= 1 else {
      throw FaceVisionException(.invalidArgument, "leftEye and rightEye must be distinct points.")
    }
    // Subject's left eye sits on the image right, so level eyes give 0.
    let angle = atan2(dy, dx) * 180 / .pi
    return try autoreleasepool {
      let url = try fileURL(from: uri)
      let image = try decodeUpright(try imageSource(url), url, maxDimension: nil)
      var result = try render(image, face: rect, rotation: -angle, options: options)
      result["rotation"] = Double(-angle)
      return result
    }
  }

  // MARK: - Crop/rotate

  /**
   Rotates the image by `rotation` degrees (clockwise on screen) around the face centre and
   crops the padded face box in that rotated space, clamped to the rotated image's bounds.
   */
  private static func render(_ image: CGImage, face: CGRect, rotation: CGFloat, options: CropOptions) throws -> [String: Any] {
    guard face.width > 0, face.height > 0, face.minX.isFinite, face.minY.isFinite else {
      throw FaceVisionException(.invalidArgument, "The face rect must have a positive size.")
    }
    guard options.padding >= 0, options.quality >= 0, options.quality <= 1 else {
      throw FaceVisionException(.invalidArgument, "padding must be >= 0 and quality in 0..1.")
    }

    let center = CGPoint(x: face.midX, y: face.midY)
    let radians = rotation * .pi / 180
    let rotate = CGAffineTransform(translationX: center.x, y: center.y)
      .rotated(by: radians)
      .translatedBy(x: -center.x, y: -center.y)
    let bounds = CGRect(x: 0, y: 0, width: image.width, height: image.height).applying(rotate)

    let padding = CGFloat(options.padding)
    var width = face.width * (1 + 2 * padding)
    var height = face.height * (1 + 2 * padding)
    if options.square {
      width = max(width, height)
      height = width
    }
    var crop = CGRect(x: center.x - width / 2, y: center.y - height / 2, width: width, height: height)
      .intersection(bounds)
    if options.square && !crop.isNull && abs(crop.width - crop.height) > 0.5 {
      let side = min(
        width,
        2 * min(center.x - bounds.minX, bounds.maxX - center.x),
        2 * min(center.y - bounds.minY, bounds.maxY - center.y)
      )
      crop = CGRect(x: center.x - side / 2, y: center.y - side / 2, width: side, height: side)
    }
    crop = CGRect(x: crop.minX.rounded(), y: crop.minY.rounded(), width: crop.width.rounded(), height: crop.height.rounded())
    guard !crop.isNull, crop.width >= 1, crop.height >= 1 else {
      throw FaceVisionException(.invalidArgument, "The face rect is outside the image.")
    }

    var outputSize = crop.size
    if let size = options.outputSize {
      guard size.width >= 1, size.height >= 1 else {
        throw FaceVisionException(.invalidArgument, "outputSize must be at least 1x1.")
      }
      outputSize = CGSize(width: size.width.rounded(), height: size.height.rounded())
    }

    let png = options.format == "png"
    let format = UIGraphicsImageRendererFormat()
    format.scale = 1
    format.opaque = !png
    let rendered = UIGraphicsImageRenderer(size: outputSize, format: format).image { context in
      let cg = context.cgContext
      cg.interpolationQuality = .high
      cg.scaleBy(x: outputSize.width / crop.width, y: outputSize.height / crop.height)
      cg.translateBy(x: -crop.minX, y: -crop.minY)
      cg.concatenate(rotate)
      UIImage(cgImage: image).draw(at: .zero)
    }
    guard let output = rendered.cgImage else {
      throw FaceVisionException(.imageWriteFailed, "Could not render the cropped face.")
    }
    return try write(output, png: png, quality: options.quality)
  }

  // MARK: - Input

  static func fileURL(from uri: String) throws -> URL {
    let url: URL
    if uri.hasPrefix("file://") {
      guard let parsed = URL(string: uri) else {
        throw FaceVisionException(.invalidUri, "'\(uri)' is not a valid file URI.")
      }
      url = parsed
    } else if uri.hasPrefix("/") {
      url = URL(fileURLWithPath: uri)
    } else {
      throw FaceVisionException(.invalidUri, "Only local file:// URIs are supported, got '\(uri)'.")
    }

    guard FileManager.default.fileExists(atPath: url.path) else {
      throw FaceVisionException(.fileNotFound, "No file exists at \(url.path).")
    }
    return url
  }

  private static func imageSource(_ url: URL) throws -> CGImageSource {
    let options = [kCGImageSourceShouldCache: false] as CFDictionary
    guard let source = CGImageSourceCreateWithURL(url as CFURL, options), CGImageSourceGetCount(source) > 0 else {
      throw FaceVisionException(.invalidImage, "Could not read an image from \(url.lastPathComponent).")
    }
    return source
  }

  /// Pixel size after the EXIF orientation is applied.
  private static func uprightSize(_ source: CGImageSource, _ url: URL) throws -> CGSize {
    guard let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
      let width = properties[kCGImagePropertyPixelWidth] as? Int,
      let height = properties[kCGImagePropertyPixelHeight] as? Int,
      width > 0, height > 0 else {
      throw FaceVisionException(.invalidImage, "\(url.lastPathComponent) is not a decodable image.")
    }
    // 5...8 are the orientations that swap the axes.
    let orientation = properties[kCGImagePropertyOrientation] as? UInt32 ?? 1
    return orientation >= 5 && orientation <= 8
      ? CGSize(width: height, height: width)
      : CGSize(width: width, height: height)
  }

  /// Bakes the EXIF transform into the pixels and applies `maxDimension` in the same pass.
  static func decodeUpright(_ source: CGImageSource, _ url: URL?, maxDimension: Int?) throws -> CGImage {
    let name = url?.lastPathComponent ?? "the image"
    guard let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil) as? [CFString: Any],
      let width = properties[kCGImagePropertyPixelWidth] as? Int,
      let height = properties[kCGImagePropertyPixelHeight] as? Int,
      width > 0, height > 0 else {
      throw FaceVisionException(.invalidImage, "\(name) is not a decodable image.")
    }

    let longestEdge = max(width, height)
    let limit = max(min(maxDimension ?? longestEdge, longestEdge), 1)
    let options: [CFString: Any] = [
      kCGImageSourceCreateThumbnailFromImageAlways: true,
      kCGImageSourceCreateThumbnailWithTransform: true,
      kCGImageSourceShouldCacheImmediately: true,
      kCGImageSourceThumbnailMaxPixelSize: limit
    ]
    guard let image = CGImageSourceCreateThumbnailAtIndex(source, 0, options as CFDictionary) else {
      throw FaceVisionException(.invalidImage, "Could not decode \(name).")
    }
    return image
  }

  // MARK: - Output

  static func write(_ image: CGImage, png: Bool, quality: Double) throws -> [String: Any] {
    do {
      try FileManager.default.createDirectory(at: outputDirectory, withIntermediateDirectories: true)
    } catch {
      throw FaceVisionException(.imageWriteFailed, "Could not create the output directory.", cause: error)
    }
    let url = outputDirectory.appendingPathComponent("\(UUID().uuidString).\(png ? "png" : "jpg")")
    let type = (png ? UTType.png : UTType.jpeg).identifier as CFString

    guard let destination = CGImageDestinationCreateWithURL(url as CFURL, type, 1, nil) else {
      throw FaceVisionException(.imageWriteFailed, "Could not create \(url.lastPathComponent).")
    }
    let properties: [CFString: Any] = png ? [:] : [kCGImageDestinationLossyCompressionQuality: quality]
    CGImageDestinationAddImage(destination, image, properties as CFDictionary)
    guard CGImageDestinationFinalize(destination) else {
      throw FaceVisionException(.imageWriteFailed, "Could not write \(url.path).")
    }
    return ["uri": url.absoluteString, "width": image.width, "height": image.height]
  }
}

File: android/src/main/java/expo/modules/facevision/FaceImageProcessor.kt
package expo.modules.facevision

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.Paint
import android.net.Uri
import android.os.SystemClock
import androidx.exifinterface.media.ExifInterface
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import kotlinx.coroutines.suspendCancellableCoroutine
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.util.UUID
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.math.atan2
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.roundToInt

internal class FaceImageProcessor(
  private val context: Context,
  cacheRoot: File
) {
  private val outputDirectory = outputDirectory(cacheRoot)

  suspend fun detectFaces(uri: String, options: DetectionOptions): Map<String, Any> {
    val start = SystemClock.elapsedRealtime()
    return withInput(uri) { file ->
      val maxSide = when (options.performanceMode) {
        "fast" -> 640
        "accurate" -> null
        else -> 1280
      }
      val (uprightWidth, uprightHeight) = uprightSize(file)
      val bitmap = decodeUpright(file, maxSide)
      try {
        val scale = uprightWidth.toFloat() / bitmap.width
        val faces = detect(bitmap, options)
        mapOf(
          "faces" to faces
            // setMinFaceSize is a hint, not a filter.
            .filter { it.boundingBox.width() * scale / uprightWidth >= options.minFaceSize }
            .map { faceToMap(it, options, scale) },
          "image" to mapOf("width" to uprightWidth, "height" to uprightHeight, "uri" to uri),
          "processingTime" to (SystemClock.elapsedRealtime() - start).toDouble()
        )
      } finally {
        bitmap.recycle()
      }
    }
  }

  fun analyzeRegion(uri: String, rect: RectRecord?): Map<String, Double> = withInput(uri) { file ->
    val bitmap = decodeUpright(file, null)
    try {
      val bounds = if (rect == null) {
        android.graphics.Rect(0, 0, bitmap.width, bitmap.height)
      } else {
        clampRect(rect.x, rect.y, rect.width, rect.height, bitmap.width, bitmap.height)
          ?: throw FaceVisionException(ErrorCodes.INVALID_ARGUMENT, "The region does not overlap the image.")
      }
      regionStats(bounds) { x, y ->
        val c = bitmap.getPixel(x, y)
        0.299 * Color.red(c) + 0.587 * Color.green(c) + 0.114 * Color.blue(c)
      }
    } finally {
      bitmap.recycle()
    }
  }

  fun cropFace(uri: String, rect: RectRecord, options: CropOptions): Map<String, Any> = withInput(uri) { file ->
    val bitmap = decodeUpright(file, null)
    try {
      render(bitmap, rect, 0.0, options)
    } finally {
      bitmap.recycle()
    }
  }

  fun alignFace(
    uri: String,
    rect: RectRecord,
    leftEye: PointRecord,
    rightEye: PointRecord,
    options: CropOptions
  ): Map<String, Any> {
    val dx = leftEye.x - rightEye.x
    val dy = leftEye.y - rightEye.y
    if (!dx.isFinite() || !dy.isFinite() || hypot(dx, dy) < 1.0) {
      throw FaceVisionException(ErrorCodes.INVALID_ARGUMENT, "The eye points are missing or too close together.")
    }
    // The subject's left eye is on the image's right, so a level face gives 0.
    val rotation = -Math.toDegrees(atan2(dy, dx))
    return withInput(uri) { file ->
      val bitmap = decodeUpright(file, null)
      try {
        render(bitmap, rect, rotation, options) + ("rotation" to rotation)
      } finally {
        bitmap.recycle()
      }
    }
  }

  /**
   * Crops the padded face box out of [source] rotated by [rotation] degrees (clockwise on screen)
   * around the face center. The crop is clamped against the source bounds, so rotated corners
   * that fall outside the source come out black (jpeg) or transparent (png).
   */
  private fun render(source: Bitmap, rect: RectRecord, rotation: Double, options: CropOptions): Map<String, Any> {
    if (!(rect.width > 0 && rect.height > 0 && rect.x.isFinite() && rect.y.isFinite())) {
      throw FaceVisionException(ErrorCodes.INVALID_ARGUMENT, "The face rect must have a positive size.")
    }
    val imageWidth = source.width.toDouble()
    val imageHeight = source.height.toDouble()
    val cx = rect.x + rect.width / 2
    val cy = rect.y + rect.height / 2
    if (cx < 0 || cy < 0 || cx > imageWidth || cy > imageHeight) {
      throw FaceVisionException(ErrorCodes.INVALID_ARGUMENT, "The face rect is outside the image.")
    }

    val padding = max(options.padding, 0.0)
    val width = rect.width * (1 + 2 * padding)
    val height = rect.height * (1 + 2 * padding)
    val left: Int
    val top: Int
    val cropWidth: Int
    val cropHeight: Int
    if (options.square) {
      // Largest square centered on the face that still fits, so the face stays centered.
      val side = minOf(max(width, height), 2 * cx, 2 * (imageWidth - cx), 2 * cy, 2 * (imageHeight - cy))
      cropWidth = max(floor(side).toInt(), 1)
      cropHeight = cropWidth
      left = (cx - cropWidth / 2.0).roundToInt().coerceIn(0, max(source.width - cropWidth, 0))
      top = (cy - cropHeight / 2.0).roundToInt().coerceIn(0, max(source.height - cropHeight, 0))
    } else {
      val bounds = clampRect(cx - width / 2, cy - height / 2, width, height, source.width, source.height)
        ?: throw FaceVisionException(ErrorCodes.INVALID_ARGUMENT, "The face rect is outside the image.")
      left = bounds.left
      top = bounds.top
      cropWidth = bounds.width()
      cropHeight = bounds.height()
    }

    val target = options.outputSize
    val outWidth = target?.width?.roundToInt()?.takeIf { it > 0 } ?: cropWidth
    val outHeight = target?.height?.roundToInt()?.takeIf { it > 0 } ?: cropHeight

    val matrix = Matrix().apply {
      setRotate(rotation.toFloat(), cx.toFloat(), cy.toFloat())
      postTranslate(-left.toFloat(), -top.toFloat())
      postScale(outWidth.toFloat() / cropWidth, outHeight.toFloat() / cropHeight)
    }
    val output = Bitmap.createBitmap(outWidth, outHeight, Bitmap.Config.ARGB_8888)
    try {
      Canvas(output).drawBitmap(source, matrix, Paint(Paint.FILTER_BITMAP_FLAG))
      return writeImage(outputDirectory, output, options.format == "png", options.quality)
    } finally {
      output.recycle()
    }
  }

  private suspend fun detect(bitmap: Bitmap, options: DetectionOptions): List<Face> {
    val detector = FaceDetection.getClient(options.toDetectorOptions())
    try {
      return suspendCancellableCoroutine { continuation ->
        detector.process(InputImage.fromBitmap(bitmap, 0))
          .addOnSuccessListener { continuation.resume(it) }
          .addOnFailureListener { error ->
            continuation.resumeWithException(
              FaceVisionException(ErrorCodes.DETECTION_FAILED, "ML Kit face detection failed: ${error.message}", error)
            )
          }
          .addOnCanceledListener { continuation.cancel() }
      }
    } finally {
      detector.close()
    }
  }

  private inline fun <T> withInput(uri: String, block: (File) -> T): T {
    val input = resolveInput(uri)
    try {
      return block(input.file)
    } finally {
      input.cleanUp()
    }
  }

  private class ResolvedInput(val file: File, private val temporary: Boolean) {
    fun cleanUp() {
      if (temporary) file.delete()
    }
  }

  /**
   * `content://` from a photo picker is not a filesystem path, and BitmapFactory plus
   * ExifInterface both need to read the source, so it lands in a temp file first.
   */
  private fun resolveInput(uri: String): ResolvedInput {
    if (uri.isBlank()) {
      throw FaceVisionException(ErrorCodes.INVALID_URI, "The image URI is empty.")
    }
    val parsed = Uri.parse(uri)

    return when (parsed.scheme) {
      null, "file" -> {
        val path = parsed.path
          ?: throw FaceVisionException(ErrorCodes.INVALID_URI, "'$uri' has no file path.")
        val file = File(path)
        if (!file.exists()) {
          throw FaceVisionException(ErrorCodes.FILE_NOT_FOUND, "No file exists at $path.")
        }
        ResolvedInput(file, temporary = false)
      }

      "content" -> ResolvedInput(copyToCache(parsed), temporary = true)

      else -> throw FaceVisionException(
        ErrorCodes.INVALID_URI,
        "Unsupported URI scheme '${parsed.scheme}'. Use file:// or content://."
      )
    }
  }

  private fun copyToCache(uri: Uri): File {
    ensureOutputDirectory(outputDirectory)
    val target = File(outputDirectory, "input-${UUID.randomUUID()}")
    try {
      val stream = context.contentResolver.openInputStream(uri)
        ?: throw FaceVisionException(ErrorCodes.FILE_NOT_FOUND, "Could not open $uri.")
      stream.use { input ->
        FileOutputStream(target).use { output -> input.copyTo(output) }
      }
    } catch (error: FileNotFoundException) {
      throw FaceVisionException(ErrorCodes.FILE_NOT_FOUND, "Could not open $uri.", error)
    } catch (error: SecurityException) {
      throw FaceVisionException(ErrorCodes.INVALID_URI, "No read permission for $uri.", error)
    } catch (error: IOException) {
      throw FaceVisionException(ErrorCodes.FILE_NOT_FOUND, "Could not read $uri.", error)
    }
    return target
  }

  private fun uprightSize(file: File): Pair<Int, Int> {
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(file.absolutePath, bounds)
    if (bounds.outWidth <= 0 || bounds.outHeight <= 0) {
      throw FaceVisionException(ErrorCodes.INVALID_IMAGE, "${file.name} is not a decodable image.")
    }
    return if (exifOrientation(file) in SWAPPED_ORIENTATIONS) {
      bounds.outHeight to bounds.outWidth
    } else {
      bounds.outWidth to bounds.outHeight
    }
  }

  /** Decodes at (or under) [maxDimension] and bakes the EXIF orientation into the pixels. */
  private fun decodeUpright(file: File, maxDimension: Int?): Bitmap {
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    BitmapFactory.decodeFile(file.absolutePath, bounds)
    if (bounds.outWidth <= 0 || bounds.outHeight <= 0) {
      throw FaceVisionException(ErrorCodes.INVALID_IMAGE, "${file.name} is not a decodable image.")
    }

    val decodeOptions = BitmapFactory.Options().apply {
      inPreferredConfig = Bitmap.Config.ARGB_8888
      inSampleSize = sampleSizeFor(bounds.outWidth, bounds.outHeight, maxDimension)
    }
    val decoded = BitmapFactory.decodeFile(file.absolutePath, decodeOptions)
      ?: throw FaceVisionException(ErrorCodes.INVALID_IMAGE, "Could not decode ${file.name}.")

    return applyExifOrientation(scaleDown(decoded, maxDimension), exifOrientation(file))
  }

  private fun sampleSizeFor(width: Int, height: Int, maxDimension: Int?): Int {
    if (maxDimension == null || maxDimension <= 0) return 1
    var sample = 1
    while (max(width, height) / (sample * 2) >= maxDimension) {
      sample *= 2
    }
    return sample
  }

  /** inSampleSize only halves, so trim the remainder to hit maxDimension exactly. */
  private fun scaleDown(bitmap: Bitmap, maxDimension: Int?): Bitmap {
    if (maxDimension == null || maxDimension <= 0) return bitmap
    val longestEdge = max(bitmap.width, bitmap.height)
    if (longestEdge <= maxDimension) return bitmap
    val ratio = maxDimension.toFloat() / longestEdge
    val scaled = Bitmap.createScaledBitmap(
      bitmap,
      max((bitmap.width * ratio).toInt(), 1),
      max((bitmap.height * ratio).toInt(), 1),
      true
    )
    if (scaled !== bitmap) bitmap.recycle()
    return scaled
  }

  private fun exifOrientation(file: File): Int = try {
    ExifInterface(file.absolutePath)
      .getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_NORMAL)
  } catch (error: IOException) {
    ExifInterface.ORIENTATION_NORMAL
  }

  private fun applyExifOrientation(bitmap: Bitmap, orientation: Int): Bitmap {
    val matrix = Matrix()
    when (orientation) {
      ExifInterface.ORIENTATION_ROTATE_90 -> matrix.postRotate(90f)
      ExifInterface.ORIENTATION_ROTATE_180 -> matrix.postRotate(180f)
      ExifInterface.ORIENTATION_ROTATE_270 -> matrix.postRotate(270f)
      ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.postScale(-1f, 1f)
      ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.postScale(1f, -1f)
      ExifInterface.ORIENTATION_TRANSPOSE -> {
        matrix.postRotate(90f)
        matrix.postScale(-1f, 1f)
      }
      ExifInterface.ORIENTATION_TRANSVERSE -> {
        matrix.postRotate(270f)
        matrix.postScale(-1f, 1f)
      }
      else -> return bitmap
    }

    val oriented = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    if (oriented !== bitmap) bitmap.recycle()
    return oriented
  }

  companion object {
    private val SWAPPED_ORIENTATIONS = setOf(
      ExifInterface.ORIENTATION_ROTATE_90,
      ExifInterface.ORIENTATION_ROTATE_270,
      ExifInterface.ORIENTATION_TRANSPOSE,
      ExifInterface.ORIENTATION_TRANSVERSE
    )

    fun outputDirectory(cacheRoot: File) = File(cacheRoot, "face-vision")

    fun ensureOutputDirectory(directory: File) {
      if (!directory.exists() && !directory.mkdirs()) {
        throw FaceVisionException(ErrorCodes.IMAGE_WRITE_FAILED, "Could not create ${directory.absolutePath}.")
      }
    }

    fun writeImage(directory: File, bitmap: Bitmap, png: Boolean, quality: Double): Map<String, Any> {
      ensureOutputDirectory(directory)
      val file = File(directory, "${UUID.randomUUID()}.${if (png) "png" else "jpg"}")
      try {
        FileOutputStream(file).use { output ->
          val format = if (png) Bitmap.CompressFormat.PNG else Bitmap.CompressFormat.JPEG
          val compressQuality = (quality.coerceIn(0.0, 1.0) * 100).roundToInt()
          if (!bitmap.compress(format, compressQuality, output)) {
            throw IOException("Bitmap.compress returned false")
          }
        }
      } catch (error: IOException) {
        throw FaceVisionException(ErrorCodes.IMAGE_WRITE_FAILED, "Could not write ${file.absolutePath}.", error)
      }
      return mapOf(
        "uri" to Uri.fromFile(file).toString(),
        "width" to bitmap.width,
        "height" to bitmap.height
      )
    }
  }
}
